### PR TITLE
Migrate axum integration to use `with_state` over `layer(Extension`

### PR DESCRIPTION
**Pull request by snapbug**
In axum, `State` provides compile-time type safety that `Extension` does not (https://github.com/tokio-rs/axum/discussions/1830).

When using `with_state` the axum router becomes `Router<_your state_>` rather than `Router`, meaning that `LeptosRoutes` is not `impl`'d for it. Update the default impl to work with a stateful router, provided the state can provide the options.

The addition of the `LeptosOptionProvider` trait allows the impl to be generic on any custom state struct, as long as it can provide the options. I added an impl for `LeptosOptions` itself, so all the examples continue to work when updated to with_state.

The `session_auth_axum` might be a good candidate to illustrate integration with a custom state, as it also uses other `Extension` layers, I'd be happy to do this 🙂

### CODE REVIEW

**Comment by gbj** [comment]
This is failing running `cargo check` on `errors_axum` with the following—I haven't taken a look at it otherwise!
```
error[E0599]: the method `leptos_routes` exists for struct `Router<Arc<LeptosOptions>>`, but its trait bounds were not satisfied
  --> src/main.rs:55:10
   |
52 |       let app = Router::new()
   |  _______________-
53 | |         .route("/api/*fn_name", post(leptos_axum::handle_server_fns))
54 | |         .route("/special/:id", get(custom_handler))
55 | |         .leptos_routes(
   | |         -^^^^^^^^^^^^^ method cannot be called on `Router<Arc<LeptosOptions>>` due to unsatisfied trait bounds
   | |_________|
   | 
   |
  ::: /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.6.18/src/routing/mod.rs:59:1
   |
59 |   pub struct Router<S = (), B = Body> {
   |   ----------------------------------- doesn't satisfy `_: LeptosRoutes<Arc<LeptosOptions>>`
  --> /rustc/cba14074bb4cc12bfe918eabd0d52a3999b2a461/library/alloc/src/sync.rs:249:1
   |
   = note: doesn't satisfy `Arc<leptos::LeptosOptions>: LeptosOptionProvider`
   |
   = note: the following trait bounds were not satisfied:
           `Arc<leptos::LeptosOptions>: LeptosOptionProvider`
           which is required by `Router<Arc<leptos::LeptosOptions>>: LeptosRoutes<Arc<leptos::LeptosOptions>>`

warning: unused import: `LeptosRoutes`
  --> src/main.rs:16:44
   |
16 |     use leptos_axum::{generate_route_list, LeptosRoutes};
   |                                            ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

For more information about this error, try `rustc --explain E0599`.
warning: `errors_axum` (bin "errors_axum") generated 1 warning
error: could not compile `errors_axum` (bin "errors_axum") due to previous error; 1 warning emitted
[cargo-make][1] ERROR - Error while executing command, exit code: 101
[cargo-make][1] WARN - Build Failed.
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
Error: Process completed with exit code 1.
```

**Comment by snapbug** [comment]
> This is failing running `cargo check` on `errors_axum` with the following—I haven't taken a look at it otherwise!

Yeah, I saw those, hence putting back to draft state, I'd been testing with my own setup rather than with the examples available (bar changing them to `State`), so clearly missed something that's different in the examples. Will mark ready when/if I can resolve 🙂

File: examples/errors_axum/Cargo.toml [hunk 1]
```diff
@@ -10,12 +10,12 @@ crate-type = ["cdylib", "rlib"]
...
-leptos = { path = "../../../leptos/leptos", default-features = false, features = [
+leptos = { path = "../../leptos", default-features = false, features = [
   "serde",
 ] }
-leptos_axum = { path = "../../../leptos/integrations/axum", default-features = false, optional = true }
-leptos_meta = { path = "../../../leptos/meta", default-features = false }
-leptos_router = { path = "../../../leptos/router", default-features = false }
+leptos_axum = { path = "../../integrations/axum", default-features = false, optional = true }
+leptos_meta = { path = "../../meta", default-features = false }
+leptos_router = { path = "../../router", default-features = false }
```
Comment (snapbug) [inline examples/errors_axum/Cargo.toml:18]: I had a checkout of leptos as a sibling to this fork ... this took a while 😆

**Comment by snapbug** [comment]
All sorted!

**Comment by snapbug** [comment]
So `State` is a type-safe replacement for `Extension` - so it should be safer than before (you could forget to add the extension layer, and it would be a run time error rather than compile time. But otherwise, it should all function exactly the same. 🙂

**Comment by benwis** [comment]
Yeah, I don't see a problem with this. Looks good to me

**Review by benwis — COMMENTED** [review]
LGTM. State is basically an Extension replacement

**Comment by gbj** [comment]
Thanks!

**Pull request merged**